File: leetcode/rust-leetcode/src/1905.count-sub-islands.rs
use std::collections::*;

type Matrix = Vec<Vec<i32>>;

static directions: [(isize, isize); 4] = [(-1, 0), (1, 0), (0, -1), (0, 1)];

impl Solution {
    pub fn count_sub_islands(grid1: Matrix, mut grid2: Matrix) -> i32 {
        let (m, n) = (grid1.len(), grid1[0].len());

        fn bfs(g1: &Matrix, g2: &mut Matrix, start: (usize, usize), m: usize, n: usize) -> i32 {
            let mut q = VecDeque::new();
            q.push_back(start);
            let (mut di, mut dj) = (0, 0);
            let mut flag = g1[start.0][start.1] == 1;
            g2[start.0][start.1] = 2;
            while let Some((i, j)) = q.pop_front() {
                for dir in directions.iter() {
                    di = (i.saturating_add_signed(dir.0)).min(m);
                    dj = (j.saturating_add_signed(dir.1)).min(n);
                    if g2[di][dj] == 1 {
                        g2[di][dj] = 2;
                        q.push_back((di, dj));
                        if flag && g1[di][dj] != 1 {
                            flag = false;
                        }
                    }
                }
            }
            flag as _
        }

        let mut ans = 0;
        for i in 0..m {
            for j in 0..n {
                if grid2[i][j] == 1 {
                    ans += bfs(&grid1, &mut grid2, (i, j), m-1, n-1);
                }
            }
        }
        ans
    }
}

File: leetcode/python/1905.count-sub-islands.py
class Solution:
    def countSubIslands(self, grid1: List[List[int]], grid2: List[List[int]]) -> int:
        dirs = [[0, 1], [0, -1], [1, 0], [-1, 0]]
        m, n = len(grid1), len(grid1[0])
        def bfs(i: int, j: int) -> int:
            isSubIsland = 1 if grid1[i][j] == 1 else 0
            q = deque([[i, j]])
            grid2[i][j] = 2
            while q:
                x, y = q.popleft()
                for dxy in dirs:
                    dx = x + dxy[0]
                    dy = y + dxy[1]
                    if dx < 0 or dy < 0 or dx == m or dy == n or grid2[dx][dy] != 1:
                        continue
                    if grid1[dx][dy] == 0:
                        isSubIsland = 0
                    grid2[dx][dy] = 2
                    q.append([dx, dy])
            return isSubIsland
        ans = 0
        for i in range(m):
            for j in range(n):
                if grid2[i][j] == 1:
                    ans += bfs(i, j)
        return ans

